### ISSUE TITLE
🤗 [Question]: Setup of SSE Fiber with fasthttp.StreamWriter - event source is pending / never connects ...

### ISSUE THREAD

**Issue by michealroberts** [gofiber/fiber#2837]
### Question Description

**Versions**:

Go 1.21.5
github.com/gofiber/fiber/v2 v2.52.0
github.com/valyala/fasthttp v1.51.0

---

**Issue**

I have the following logic inside of an SSE handler:

```go
// Peak at the incoming Accept request header:
accept := c.Request().Header.Peek("Accept")

// Check whether the Accept header is set to text/event-stream:
if c.Accepts("text/event-stream") == "text/event-stream" && strings.Contains(string(accept), "text/event-stream") {
  ctx := c.Context()

  ctx.SetContentType("text/event-stream")

  ctx.Response.Header.Set("Cache-Control", "no-cache")
  ctx.Response.Header.Set("Connection", "keep-alive")
  ctx.Response.Header.Set("Transfer-Encoding", "chunked")
  ctx.Response.Header.Set("Access-Control-Allow-Headers", "Cache-Control")
  ctx.Response.Header.Set("Access-Control-Allow-Credentials", "true")

  ctx.SetBodyStreamWriter(fasthttp.StreamWriter(func(w *bufio.Writer) {
    defer func() {
      if r := recover(); r != nil {
        fmt.Println("Recovered in SSE writer:", r)
      }
    }()

    ticker := time.NewTicker(1 * time.Second)
    defer ticker.Stop()

    for {
      select {
      case <-ticker.C:
        status, err := GetStatus(telescope)

        if err != nil {
          fmt.Printf("Error while getting status: %v\n", err)
          return
        }

        data, err := json.Marshal(status)

        if err != nil {
          fmt.Printf("Error while marshaling JSON: %v\n", err)
          return
        }

        fmt.Fprintf(w, "data: %s\n\n", string(data))

        fmt.Println(string(data))

        if err := w.Flush(); err != nil {
          fmt.Printf("Error while flushing: %v. Closing connection.\n", err)
          return
        }
      case <-c.Context().Done():
        fmt.Println("Client disconnected. Closing connection.")
        return
      }
    }
  }))

  return nil
}
```

Which to me, looks good. I can log the message without issue ...

However, when connecting from a browser side client ... the connection is stuck in the "connecting" phase of the event source connection.

I can see, that when the endpoint is called from my client, the server is logging correctly:

![CleanShot 2024-02-08 at 11 50 20](https://github.com/valyala/fasthttp/assets/84131395/e0e964a1-f4ed-4141-ab7e-cb48baf33dc0)

But the connection remains as pending:

<img width="860" alt="CleanShot 2024-02-08 at 11 49 24@2x" src="https://github.com/valyala/fasthttp/assets/84131395/bc83f9f9-64f4-4a6b-b25b-2b91a566773a">

Also, no errors when requesting `application/json` (so for me, it isn't a case that the underlying code has an issue):

<img width="538" alt="CleanShot 2024-02-08 at 12 03 50@2x" src="https://github.com/gofiber/fiber/assets/84131395/9048a710-b0c2-4f61-a087-a20dc1e4683a">

The front end JS code is standard for the EventSource API.

--- 

**Headers**

<img width="744" alt="CleanShot 2024-02-08 at 11 54 39@2x" src="https://github.com/valyala/fasthttp/assets/84131395/d85fb4de-9638-418e-a0b5-5cf33b807dd4">

---

**Reproduction**

I can also provide access to the repository for a minimal reproduction if the issue isn't apparent from what I have supplied if needed, please just request access for your username and I can provide it (as long as you are listed as a core maintainer of this repo).

### Code Snippet (optional)

```go
/*****************************************************************************************************************/

//	@author		Michael Roberts <michael@observerly.com>
//	@package	@observerly/nox/telescope
//	@license	Copyright © 2021-2023 observerly

/*****************************************************************************************************************/

package telescope

/*****************************************************************************************************************/

import (
	"bufio"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/observerly/alpacago/pkg/alpacago"
	"github.com/valyala/fasthttp"

	"nox/internal/common"
	"nox/internal/middleware"
)

/*****************************************************************************************************************/

type GetStatusHandlerResponse struct {
	Connected bool `json:"connected"`
}

/*****************************************************************************************************************/

type GetStatusChannels struct {
	Connected chan bool `json:"connected"`
}

/*****************************************************************************************************************/

func GetStatus(telescope *alpacago.Telescope) (GetStatusHandlerResponse, error) {
	// Create channels for the status values:
	channels := GetStatusChannels{}

	// Create a wait group for the status values:
	wg, channels, errors := common.SetupWaitGroupForStruct(channels)

	// Get the connection status:
	go func() {
		defer wg.Done()
		common.RetrieveAndSendToChannel(telescope.IsConnected, channels.Connected, errors)
	}()

	go func() {
		// Wait for all the goroutines to finish:
		wg.Wait()
		// Close the channels:
		common.CloseChannelsForStruct(channels)
	}()

	status := &GetStatusHandlerResponse{}

	// Extract the values from the channels:
	err := common.ExtractValueFromChannelStruct(channels, status)

	// Check if we encountered any errors while extracting the values:
	if len(errors) > 0 {
		return *status, fmt.Errorf("encountered errors while retrieving status values: %v", errors)
	}

	// If we encounter an error, return the error:
	if err != nil {
		return *status, err
	}

	// Return the status values:
	return *status, nil
}

/*****************************************************************************************************************/

func GetStatusHandler(c *fiber.Ctx) error {
	// Get the telescope client from the context middleware:
	telescope, ok := c.Locals("telescope").(*alpacago.Telescope)

	// If the telescope client is not available in the context, return an error:
	if !ok {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Telescope is not available in context",
		})
	}

	// Peak at the incoming Accept request header:
	accept := c.Request().Header.Peek("Accept")

	// Check whether the Accept header is set to text/event-stream:
	if c.Accepts("text/event-stream") == "text/event-stream" && strings.Contains(string(accept), "text/event-stream") {
		ctx := middleware.TextEventStream(c)

		ctx.SetBodyStreamWriter(fasthttp.StreamWriter(func(w *bufio.Writer) {
			defer func() {
				if r := recover(); r != nil {
					fmt.Println("Recovered in SSE writer:", r)
				}
			}()

			ticker := time.NewTicker(1 * time.Second)
			defer ticker.Stop()

			for {
				select {
				case <-ticker.C:
					status, err := GetStatus(telescope)

					if err != nil {
						fmt.Printf("Error while getting status: %v\n", err)
						return
					}

					data, err := json.Marshal(status)

					if err != nil {
						fmt.Printf("Error while marshaling JSON: %v\n", err)
						return
					}

					fmt.Fprintf(w, "data: %s\n\n", string(data))

					fmt.Println(string(data))

					if err := w.Flush(); err != nil {
						fmt.Printf("Error while flushing: %v. Closing connection.\n", err)
						return
					}
				case <-c.Context().Done():
					fmt.Println("Client disconnected. Closing connection.")
					return
				}
			}
		}))

		return nil
	}

	// Get the telescope status
	status, err := GetStatus(telescope)

	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(
			common.ErrorResponse{
				Error: err.Error(),
			},
		)
	}

	// Return the telescope status:
	return c.JSON(status)
}

/*****************************************************************************************************************/
```

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have checked for existing issues that describe my questions prior to opening this one.
- [X] I understand that improperly formatted questions may be closed without explanation.

**Comment by efectn** [comment]
@michealroberts is your endpoint working if you remove
```
case <-c.Context().Done():
    fmt.Println("Client disconnected. Closing connection.")
    return
```

**Comment by michealroberts** [comment]
@efectn Yeh, so it still doesn't work when removing that unfortunately 😞 

I have a branch here: https://github.com/observerly/nox/pull/48 for a full reproduction, I've added the basic example given in the gofiber/examples repo for SSE setup, line for line, and unfortunately it doesn't work ...

Are you able to work with SSE on the latest versions of Fiber and fasthttp?

**Comment by efectn** [comment]
> @efectn Yeh, so it still doesn't work when removing that unfortunately 😞
> 
> I have a branch here: [observerly/nox#48](https://github.com/observerly/nox/pull/48) for a full reproduction that you should be able to access. In that code, I've added the basic example given in the gofiber/examples repo for SSE setup, line for line, and unfortunately it still doesn't work ...
> 
> Are you able to work with SSE on the latest versions of Fiber and fasthttp?

This one works for me https://paste.laravel.io/8c6a1464-4f52-46c1-b362-ab49f5ad60cf

![2024-02-09_18-40](https://github.com/gofiber/fiber/assets/45270788/9f1afe47-98d9-47df-af8c-6f12189b058b)

**Comment by efectn** [comment]
> @efectn Aye aye aye, ok. I feel like I have narrowed it down to be able to replicate it.
> 
> I have the ETag middleware installed from [github.com/gofiber/fiber/v2/middleware/etag](github.com/gofiber/fiber/v2/middleware/etag) ... I guess this somehow causes issues in terms of headers 🤔
> 
> What are your thoughts on this? I should be able to get a minimal reproduction.

Yes it seems. You can disable etag for specific path like:
```go
app.Use(etag.New(etag.Config{
	Next: func(c *fiber.Ctx) bool {
		return c.Path() == "/sse"
	},
}))
```

**Comment by michealroberts** [comment]
@efectn I think I will disable ETags globally for now, I have quite a number of SSE routes.

I wonder if I should open up a separate minimal, reproducible, example of the SSE + ETag issue, and maybe start to work on a possible fix ...

**Comment by sdaduanbilei** [comment]
I have the same problem.

**Comment by michealroberts** [comment]
I will start working on this issue this week.

**Comment by SillyZir** [comment]
SendStreamWriter is a thin delegation to fasthttp’s Response.SetBodyStreamWriter — Fiber adds no header or flushing logic of its own (res.go:1256-1261).

The stream writer runs after the handler has returned, but the handler releases its Ctx back into the pool via defer as soon as the handler stack returns, and a pooled ctx is reset for reuse — so the writer’s lifetime overlaps a recycled context unless something prevents it (router.go:740-746, ctx_interface.go:127-150).

That’s exactly what v3’s Abandon() exists for: it lets “the timeout and SSE middlewares” return while a goroutine keeps using the context safely, and abandoned contexts are never pooled (ctx.go:843-856). Fiber’s own v3 SSE middleware captures c.Context() and the Last-Event-ID header before starting the stream writer and calls c.Abandon() first — it deliberately avoids lazy ctx access inside the writer, which is what the snippet in this issue does with c.Context().Done() (middleware/sse/sse.go:36-41).

So the most likely culprit — not yet verified on v2.52.0 — is a use-after-release: v2 has no Abandon mechanism, so returning from the handler pools and resets the Ctx while the writer goroutine still reads it, and <-c.Context().Done() inside the writer may be waiting on a recycled request’s context. A minimal v2 repro under go run -race should surface races on ctx fields the first time the writer ticks after the handler returns.

The “pending / never connects” symptom could also be buffering between fasthttp and the browser — a dev-server proxy, nginx without X-Accel-Buffering: no, or a compression layer; v3’s SSE middleware sets that header explicitly for this reason (middleware/sse/sse.go:31-34). Quick differential: curl -N http://host:port/endpoint -H 'Accept: text/event-stream' straight at the Go process — if curl streams but the browser stalls, the intermediary is at fault.

A workaround consistent with the above: capture anything you need from the ctx before calling SendStreamWriter, and detect disconnects via w.Flush() errors — the documented contract is that Flush returns an error once the client is gone (middleware/sse/sse.go:36-41, docs/api/ctx.md:2994-3002).

**Comment by SillyZir** [comment]
Following up on my previous comment with the verification it promised. The use-after-release
was a hypothesis there; under `-race` it is now measured — though what reproduces is the
**mechanism**, under synthetic concurrency, not your browser "stuck connecting" symptom.

Pinning your exact versions (fiber v2.52.0, fasthttp v1.51.0; Go 1.25.6 here rather than your
1.21.5), a handler that registers a stream writer and then returns, with a lazy
`<-c.Context().Done()` *inside* the writer — the same shape as your snippet — reports a data
race on the first tick after the handler returns:

```
WARNING: DATA RACE
Read at 0x00c0001a42a0 by goroutine 15:
  github.com/gofiber/fiber/v2.(*Ctx).Context()
      .../fiber/v2@v2.52.0/ctx.go:456
  main.main.func1.1()
      .../main.go:25            <- <-c.Context().Done() inside the writer

Previous write at 0x00c0001a42a0 by goroutine 12:
  github.com/gofiber/fiber/v2.(*App).ReleaseCtx()
      .../fiber/v2@v2.52.0/ctx.go:193
  github.com/gofiber/fiber/v2.(*App).handler.deferwrap1()
      .../fiber/v2@v2.52.0/router.go:163
```

So: the handler's deferred `ReleaseCtx` writes the Ctx back to the pool while the writer
goroutine is still reading it. On this machine it then goes further and panics —
`invalid memory address or nil pointer dereference` in
`fasthttp.(*RequestCtx).Done()` (`server.go:2726`) — because the recycled Ctx's `fasthttp.RequestCtx`
pointer has been reset out from under the writer. 3/3 runs reproduce the race; the panic is the
same run continuing.

What this does and does not show:

- It shows `c.Context()` inside a v2 stream writer reads a pooled, already-released Ctx. That is
  a real race in the pattern, independent of any browser.
- It does **not** show this is why your `EventSource` hangs in "connecting". I forced the pool to
  recycle with 60 follow-on requests; a single idle stream may never hit it. The buffering
  differential from my earlier comment (`curl -N` straight at the Go process) is still the way to
  separate the two.
- The workaround is unchanged and is what v3's SSE middleware does: capture everything you need
  from the ctx *before* `SetBodyStreamWriter`, and detect disconnect via the `w.Flush()` error
  rather than `c.Context().Done()`.

Minimal reproduction — `go mod init`, these two files, `go run -race .`:

`go.mod`
```go
require (
	github.com/gofiber/fiber/v2 v2.52.0
	github.com/valyala/fasthttp v1.51.0
)
```

`main.go`
```go
package main

import (
	"bufio"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/valyala/fasthttp"
)

func main() {
	app := fiber.New(fiber.Config{DisableStartupMessage: true})
	app.Get("/sse", func(c *fiber.Ctx) error {
		c.Set("Content-Type", "text/event-stream")
		c.Context().SetBodyStreamWriter(fasthttp.StreamWriter(func(w *bufio.Writer) {
			for i := 0; i < 40; i++ {
				select {
				case <-c.Context().Done(): // lazy ctx access AFTER the handler returned
					return
				default:
				}
				fmt.Fprintf(w, "data: tick %d\n\n", i)
				if err := w.Flush(); err != nil {
					return
				}
				time.Sleep(25 * time.Millisecond)
			}
		}))
		return nil // handler returns; Ctx goes back to the pool
	})
	go func() { _ = app.Listen("127.0.0.1:3123") }()
	time.Sleep(300 * time.Millisecond)

	// first stream: the writer keeps ticking after we walk away
	if r, err := http.Get("http://127.0.0.1:3123/sse"); err == nil {
		_, _ = io.ReadAll(io.LimitReader(r.Body, 32))
		r.Body.Close()
	}
	// hammer more requests so the pooled ctx is recycled under the writer
	for i := 0; i < 60; i++ {
		if r, err := http.Get("http://127.0.0.1:3123/sse"); err == nil {
			_, _ = io.ReadAll(io.LimitReader(r.Body, 16))
			r.Body.Close()
		}
	}
	time.Sleep(1500 * time.Millisecond)
	fmt.Println("done")
}
```

**Comment by gaby** [comment]
@SillyZir The fix/feature was only added for fiber v3. V2 is in maintenance mode, only security fixes.
